react/features/base/avatar/components/AbstractAvatar.js

```diff
@@ -0,0 +1,177 @@
+// @flow
+
+import { PureComponent } from 'react';
+
+import { getParticipantById } from '../../participants';
+
+import { getAvatarColor, getInitials } from '../functions';
+
+export type Props = {
+
+    /**
+     * The string we base the initials on (this is generated from a list of precendences).
+     */
+    _initialsBase: ?string,
+
+    /**
+     * An URL that we validated that it can be loaded.
+     */
+    _loadableUrl: ?string,
+
+    /**
+     * A string to override the initials to generate a color of. This is handy if you don't want to make
+     * the background color match the string that the initials are generated from.
+     */
+    colorBase?: string,
+
+    /**
+     * Display name of the entity to render an avatar for (if any). This is handy when we need
+     * an avatar for a non-participasnt entity (e.g. a recent list item).
+     */
+    displayName?: string,
+
+    /**
+     * The ID of the participant to render an avatar for (if it's a participant avatar).
+     */
+    participantId?: string,
+
+    /**
+     * The size of the avatar.
+     */
+    size: number,
+
+    /**
+     * URI of the avatar, if any.
+     */
+    uri: ?string,
+}
+
+type State = {
+    avatarFailed: boolean
+}
+
+export const DEFAULT_SIZE = 65;
+
+/**
+ * Implements an abstract class to render avatars in the app.
+ */
+export default class AbstractAvatar<P: Props> extends PureComponent<P, State> {
+    /**
+     * Instantiates a new {@code Component}.
+     *
+     * @inheritdoc
+     */
+    constructor(props: P) {
+        super(props);
+
+        this.state = {
+            avatarFailed: false
+        };
+
+        this._onAvatarLoadError = this._onAvatarLoadError.bind(this);
+    }
+
+    /**
+     * Implements {@code Component#componentDidUpdate}.
+     *
+     * @inheritdoc
+     */
+    componentDidUpdate(prevProps: P) {
+        if (prevProps.uri !== this.props.uri) {
+
+            // URI changed, so we need to try to fetch it again.
+            // Eslint doesn't like this statement, but based on the React doc, it's safe if it's
+            // wrapped in a condition: https://reactjs.org/docs/react-component.html#componentdidupdate
+
+            // eslint-disable-next-line react/no-did-update-set-state
+            this.setState({
+                avatarFailed: false
+            });
```

Comment:
What if it it started as failed and a callback was called on both success and failed to set the state, so something like avatarLoaded instead of avatarFailed? That would remove setting state in _onAvatarLoadError.